### DISCUSSION TITLE
Graphcache is not updated on `cache.updateQuery` and stays empty

### DISCUSSION THREAD

**Discussion by konsti** [urql-graphql/urql#2754] [Q&A]
Hi.

We develop an application that displays speech-to-text transcription results to multiple users. Those transcripts are pushed to the front-end application via Graphql Subscriptions. There are two different use cases:
- **A transcript record is being updated** (e.g., when the current speaker is still talking, we update his sentence)
- **A new transcript record is published** (e.g., a new speaker starts speaking, or the sentence is finished, and we start a new sentence)

In our `cacheExchange`, we have defined a `UpdateResolver`:

```typescript
const transcriptCueUpdateResolver: UpdateResolver<
  TranscriptCueUpdateSubscription,
  TranscriptCueUpdateSubscriptionVariables
> = (result, args, cache) => {
  cache.updateQuery<TranscriptCuesQuery>(
    { query: TranscriptCuesDocument, variables: { recordingId: args.recordingId } },
    (data) => {
      if (!data) return null;

      const index = data.transcriptCues.findIndex(
        (entry) => entry?.id === result.transcriptCueUpdate.id,
      );
      index >= 0
        ? data.transcriptCues[index] = result.transcriptCueUpdate
        : data.transcriptCues.push(result.transcriptCueUpdate);
      return data;
    },
  );
};

cacheExchange({
  schema,
  updates: {
    Subscription: {
      transcriptCueUpdate: transcriptCueUpdateResolver,
      [...]
    },
  },
})
```

When testing this in **three** individual browser windows, **two** browser windows are updated correctly, and the user can see new `transcriptCues` and existing ones updated. However, in **the last** browser, the `data` result form the `query: TranscriptCuesDocument` is always an empty array (`data.transcriptCues = []`)!
In this browser, the user canonly see the newest transcript, as only the newest `result` is being pushed into the cache.

I can't explain this behavior and don't understand why the cache is not being updated in that browser instance. If I refresh the window, the original `Query` receives all the existing transcripts from the DB, and it works **correctly** on all browsers. The issue only appears when starting from an empty array of transcripts.

Do you have any pointers or ideas for me on how to debug and hopefully resolve this issue?

**Comment by JoviDeCroock** [comment]
I can't really tell why this would happen but the first thought I had was that update operations happen automatically hence don't need a custom updater. Have you tried without a custom updater? Becaue if that doesn't work you most likely have an unexpected typename/... going on

EDIT: I just realised you are trying to push, in that case I don't see why that would be, mind trying to do it immutably? Because the only other reason would be a missing field which isn't possible if it works in other browsers

**Reply by konsti** [reply]
@JoviDeCroock, thanks for your fast reply!

Treating `data` as immutable and returning a new modified `data` object didn't work either.

No other `Query` in our frontend touches `transcriptCues`. Could it still happen that a query wipes the cache on this browser instance?

Any ideas on how to debug this? Can I `console.log` cache actions somehow?

**Reply by JoviDeCroock** [reply]
I mean, the positive about our cache only being manipulated by GraphQL operations is that it shouldn't happen outside of your action. Do you have a way to reproduce this and what browser I should use to see it?
